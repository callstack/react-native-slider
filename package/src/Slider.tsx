import React, {useState} from 'react';
import {
  Image,
  Platform,
  StyleSheet,
  AccessibilityActionEvent,
  ViewProps,
  ViewStyle,
  ColorValue,
  NativeSyntheticEvent,
  StyleProp,
  View,
} from 'react-native';
import RCTSliderNativeComponent from './index';
//@ts-ignore
import type {ImageSource} from 'react-native/Libraries/Image/ImageSource';

import type {FC, Ref} from 'react';
import {MarkerProps} from './components/TrackMark';
import {StepsIndicator} from './components/StepsIndicator';
import {styles} from './utils/styles';
import {constants} from './utils/constants';

type Event = NativeSyntheticEvent<
  Readonly<{
    value: number;
    /**
     * Android Only.
     */
    fromUser?: boolean;
  }>
>;

type WindowsProps = Readonly<{
  /**
   * If true the slider will be inverted.
   * Default value is false.
   */
  vertical?: boolean;
}>;

type IOSProps = Readonly<{
  /**
   * Assigns a single image for the track. Only static images are supported.
   * The center pixel of the image will be stretched to fill the track.
   */
  trackImage?: ImageSource;

  /**
   * Assigns a minimum track image. Only static images are supported. The
   * rightmost pixel of the image will be stretched to fill the track.
   */
  minimumTrackImage?: ImageSource;

  /**
   * Assigns a maximum track image. Only static images are supported. The
   * leftmost pixel of the image will be stretched to fill the track.
   */
  maximumTrackImage?: ImageSource;

  /**
   * Permits tapping on the slider track to set the thumb position.
   * Defaults to false on iOS. No effect on Android or Windows.
   */
  tapToSeek?: boolean;
}>;

type Props = ViewProps &
  IOSProps &
  WindowsProps &
  Readonly<{
    /**
     * Used to style and layout the `Slider`.  See `StyleSheet.js` and
     * `DeprecatedViewStylePropTypes.js` for more info.
     */
    style?: StyleProp<ViewStyle>;

    /**
     * Write-only property representing the value of the slider.
     * Can be used to programmatically control the position of the thumb.
     * Entered once at the beginning still acts as an initial value.
     * The value should be between minimumValue and maximumValue,
     * which default to 0 and 1 respectively.
     * Default value is 0.
     *
     * This is not a controlled component, you don't need to update the
     * value during dragging.
     */
    value?: number;

    /**
     * Step value of the slider. The value should be
     * between 0 and (maximumValue - minimumValue).
     * Default value is 0.
     */
    step?: number;

    /**
     * Initial minimum value of the slider. Default value is 0.
     */
    minimumValue?: number;

    /**
     * Initial maximum value of the slider. Default value is 1.
     */
    maximumValue?: number;

    /**
     * The lower limit value of the slider. The user won't be able to slide below this limit.
     */
    lowerLimit?: number;

    /**
     * The upper limit value of the slider. The user won't be able to slide above this limit.
     */
    upperLimit?: number;

    /**
     * The color used for the track to the left of the button.
     * Overrides the default blue gradient image on iOS.
     */
    minimumTrackTintColor?: ColorValue;

    /**
     * The color used for the track to the right of the button.
     * Overrides the default blue gradient image on iOS.
     */
    maximumTrackTintColor?: ColorValue;
    /**
     * The color used to tint the default thumb images on iOS, or the
     * color of the foreground switch grip on Android.
     */
    thumbTintColor?: ColorValue;

    /**
     * If true the user won't be able to move the slider.
     * Default value is false.
     */
    disabled?: boolean;

    /**
     * Callback continuously called while the user is dragging the slider.
     */
    onValueChange?: (value: number) => void;

    /**
     * Callback that is called when the user touches the slider,
     * regardless if the value has changed. The current value is passed
     * as an argument to the callback handler.
     */
    onSlidingStart?: (value: number) => void;

    /**
     * Callback that is called when the user releases the slider,
     * regardless if the value has changed. The current value is passed
     * as an argument to the callback handler.
     */
    onSlidingComplete?: (value: number) => void;

    /**
     * Used to locate this view in UI automation tests.
     */
    testID?: string;

    /**
     * Sets an image for the thumb. Only static images are supported.
     */
    thumbImage?: ImageSource;

    /**
     * If true the slider will be inverted.
     * Default value is false.
     */
    inverted?: boolean;

    /**
     * Component to be rendered for each step indicator.
     */
    StepMarker?: FC<MarkerProps>;

    /**
     *
     */
    renderStepNumber?: boolean;

    /**
     * A string of one or more words to be announced by the screen reader.
     * Otherwise, it will announce the value as a percentage.
     * Requires passing a value to `accessibilityIncrements` to work correctly.
     * Should be a plural word, as singular units will be handled.
     */
    accessibilityUnits?: string;

    /**
     * An array of values that represent the different increments displayed
     * by the slider. All the values passed into this prop must be strings.
     * Requires passing a value to `accessibilityUnits` to work correctly.
     * The number of elements must be the same as `maximumValue`.
     */
    accessibilityIncrements?: Array<string>;
  }>;

const SliderComponent = (
  props: Props,
  forwardedRef?: Ref<typeof RCTSliderNativeComponent>,
) => {
  const {
    onValueChange,
    onSlidingStart,
    onSlidingComplete,
    onAccessibilityAction,
    ...localProps
  } = props;
  const [currentValue, setCurrentValue] = useState(
    props.value ?? props.minimumValue,
  );
  const [width, setWidth] = useState(0);

  const options = Array.from(
    {
      length:
        (localProps.maximumValue! - localProps.minimumValue!) /
          (localProps.step
            ? localProps.step
            : constants.DEFAULT_STEP_RESOLUTION) +
        1,
    },
    (_, index) => index,
  );

  const defaultStyle =
    Platform.OS === 'ios' ? styles.defaultSlideriOS : styles.defaultSlider;
  const sliderStyle = {zIndex: 1, width: width};
  const style = StyleSheet.compose(props.style, defaultStyle);

  const onValueChangeEvent = (event: Event) => {
    onValueChange && onValueChange(event.nativeEvent.value);
    setCurrentValue(event.nativeEvent.value);
  };

  const _disabled =
    typeof props.disabled === 'boolean'
      ? props.disabled
      : props.accessibilityState?.disabled === true;

  const _accessibilityState =
    typeof props.disabled === 'boolean'
      ? {...props.accessibilityState, disabled: props.disabled}
      : props.accessibilityState;

  const onSlidingStartEvent = onSlidingStart
    ? (event: Event) => {
        onSlidingStart(event.nativeEvent.value);
      }
    : null;
  const onSlidingCompleteEvent = onSlidingComplete
    ? (event: Event) => {
        onSlidingComplete(event.nativeEvent.value);
      }
    : null;
  const onAccessibilityActionEvent = onAccessibilityAction
    ? (event: AccessibilityActionEvent) => {
        onAccessibilityAction(event);
      }
    : null;

  const value =
    Number.isNaN(props.value) || !props.value ? undefined : props.value;

  const lowerLimit =
    !!localProps.lowerLimit || localProps.lowerLimit === 0
      ? localProps.lowerLimit
      : Platform.select({
          web: localProps.minimumValue,
          default: constants.LIMIT_MIN_VALUE,
        });

  const upperLimit =
    !!localProps.upperLimit || localProps.upperLimit === 0
      ? localProps.upperLimit
      : Platform.select({
          web: localProps.maximumValue,
          default: constants.LIMIT_MAX_VALUE,
        });

  return (
    <View
      onLayout={(event) => {
        setWidth(event.nativeEvent.layout.width);
      }}
      style={[styles, style, {justifyContent: 'center'}]}>
      {props.StepMarker || !!props.renderStepNumber ? (
        <StepsIndicator
          options={options}
          sliderWidth={width}
          currentValue={currentValue}
          renderStepNumber={localProps.renderStepNumber}
          thumbImage={localProps.thumbImage}
          StepMarker={localProps.StepMarker}
          isLTR={localProps.inverted}
        />
      ) : null}
      <RCTSliderNativeComponent
        {...localProps}
        value={value}
        lowerLimit={lowerLimit}
        upperLimit={upperLimit}
        accessibilityState={_accessibilityState}
        thumbImage={
          Platform.OS === 'web'
            ? props.thumbImage
            : props.StepMarker
            ? undefined
            : Image.resolveAssetSource(props.thumbImage)
        }
        ref={forwardedRef}
        style={[
          sliderStyle,
          defaultStyle,
          {alignContent: 'center', alignItems: 'center'},
        ]}
        onChange={onValueChangeEvent}
        onRNCSliderSlidingStart={onSlidingStartEvent}
        onRNCSliderSlidingComplete={onSlidingCompleteEvent}
        onRNCSliderValueChange={onValueChangeEvent}
        disabled={_disabled}
        onStartShouldSetResponder={() => true}
        onResponderTerminationRequest={() => false}
        onRNCSliderAccessibilityAction={onAccessibilityActionEvent}
        thumbTintColor={
          props.thumbImage && !!props.StepMarker
            ? 'transparent'
            : props.thumbTintColor
        }
      />
    </View>
  );
};

const SliderWithRef = React.forwardRef(SliderComponent);

SliderWithRef.defaultProps = {
  value: 0,
  minimumValue: 0,
  maximumValue: 1,
  step: 0,
  inverted: false,
  tapToSeek: false,
  lowerLimit: Platform.select({
    web: undefined,
    default: constants.LIMIT_MIN_VALUE,
  }),
  upperLimit: Platform.select({
    web: undefined,
    default: constants.LIMIT_MAX_VALUE,
  }),
};

export default SliderWithRef;
