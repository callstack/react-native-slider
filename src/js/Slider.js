/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow
 */

'use strict';

import React from 'react';
import {Image, Platform, StyleSheet} from 'react-native';
import RCTSliderNativeComponent from './RNCSliderNativeComponent';

import type {Ref} from 'react';
import type {NativeComponent} from 'react-native/Libraries/Renderer/shims/ReactNative';
import type {ImageSource} from 'react-native/Libraries/Image/ImageSource';
import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';
import type {ColorValue} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';
import type {ViewProps} from 'react-native/Libraries/Components/View/ViewPropTypes';
import type {SyntheticEvent} from 'react-native/Libraries/Types/CoreEventTypes';

type Event = SyntheticEvent<
  $ReadOnly<{|
    value: number,
    /**
     * Android Only.
     */
    fromUser?: boolean,
  |}>,
>;

type WindowsProps = $ReadOnly<{|
  /**
   * If true the slider will be inverted.
   * Default value is false.
   */
  vertical?: ?boolean,
|}>;

type IOSProps = $ReadOnly<{|
  /**
   * Assigns a single image for the track. Only static images are supported.
   * The center pixel of the image will be stretched to fill the track.
   */
  trackImage?: ?ImageSource,

  /**
   * Assigns a minimum track image. Only static images are supported. The
   * rightmost pixel of the image will be stretched to fill the track.
   */
  minimumTrackImage?: ?ImageSource,

  /**
   * Assigns a maximum track image. Only static images are supported. The
   * leftmost pixel of the image will be stretched to fill the track.
   */
  maximumTrackImage?: ?ImageSource,
|}>;

type Props = $ReadOnly<{|
  ...ViewProps,
  ...IOSProps,
  ...WindowsProps,

  /**
   * Used to style and layout the `Slider`.  See `StyleSheet.js` and
   * `DeprecatedViewStylePropTypes.js` for more info.
   */
  style?: ?ViewStyleProp,

  /**
   * Initial value of the slider. The value should be between minimumValue
   * and maximumValue, which default to 0 and 1 respectively.
   * Default value is 0.
   *
   * *This is not a controlled component*, you don't need to update the
   * value during dragging.
   */
  value?: ?number,

  /**
   * Step value of the slider. The value should be
   * between 0 and (maximumValue - minimumValue).
   * Default value is 0.
   */
  step?: ?number,

  /**
   * Initial minimum value of the slider. Default value is 0.
   */
  minimumValue?: ?number,

  /**
   * Initial maximum value of the slider. Default value is 1.
   */
  maximumValue?: ?number,

  /**
   * The color used for the track to the left of the button.
   * Overrides the default blue gradient image on iOS.
   */
  minimumTrackTintColor?: ?ColorValue,

  /**
   * The color used for the track to the right of the button.
   * Overrides the default blue gradient image on iOS.
   */
  maximumTrackTintColor?: ?ColorValue,
  /**
   * The color used to tint the default thumb images on iOS, or the
   * color of the foreground switch grip on Android.
   */
  thumbTintColor?: ?ColorValue,

  /**
   * If true the user won't be able to move the slider.
   * Default value is false.
   */
  disabled?: ?boolean,

  /**
   * Callback continuously called while the user is dragging the slider.
   */
  onValueChange?: ?(value: number) => void,

  /**
   * Callback that is called when the user touches the slider,
   * regardless if the value has changed. The current value is passed
   * as an argument to the callback handler.
   */

  onSlidingStart?: ?(value: number) => void,

  /**
   * Callback that is called when the user releases the slider,
   * regardless if the value has changed. The current value is passed
   * as an argument to the callback handler.
   */
  onSlidingComplete?: ?(value: number) => void,

  /**
   * Used to locate this view in UI automation tests.
   */
  testID?: ?string,

  /**
   * Sets an image for the thumb. Only static images are supported.
   */
  thumbImage?: ?ImageSource,

  /**
   * If true the slider will be inverted.
   * Default value is false.
   */
  inverted?: ?boolean,

  /**
   * A string of one or more words to be announced by the screen reader.
   * Otherwise, it will announce the value as a percentage.
   * Requires passing a value to `accessibilityIncrements` to work correctly.
   * Should be a plural word, as singular units will be handled.
   */
  accessibilityUnits?: string,

  /**
   * An array of values that represent the different increments displayed
   * by the slider. All the values passed into this prop must be strings.
   * Requires passing a value to `accessibilityUnits` to work correctly.
   * The number of elements must be the same as `maximumValue`.
   */
  accessibilityIncrements?: Array<string>,
|}>;

/**
 * A component used to select a single value from a range of values.
 *
 * ### Usage
 *
 * The example below shows how to use `Slider` to change
 * a value used by `Text`. The value is stored using
 * the state of the root component (`App`). The same component
 * subscribes to the `onValueChange`  of `Slider` and changes
 * the value using `setState`.
 *
 *```
 * import React from 'react';
 * import { StyleSheet, Text, View, Slider } from 'react-native';
 *
 * export default class App extends React.Component {
 *   constructor(props) {
 *     super(props);
 *     this.state = {
 *       value: 50
 *     }
 *   }
 *
 *   change(value) {
 *     this.setState(() => {
 *       return {
 *         value: parseFloat(value)
 *       };
 *     });
 *   }
 *
 *   render() {
 *     const {value} = this.state;
 *     return (
 *       <View style={styles.container}>
 *         <Text style={styles.text}>{String(value)}</Text>
 *         <Slider
 *           step={1}
 *           maximumValue={100}
 *           onValueChange={this.change.bind(this)}
 *           value={value} />
 *       </View>
 *     );
 *   }
 * }
 *
 * const styles = StyleSheet.create({
 *   container: {
 *     flex: 1,
 *     flexDirection: 'column',
 *     justifyContent: 'center'
 *   },
 *   text: {
 *     fontSize: 50,
 *     textAlign: 'center'
 *   }
 * });
 *```
 *
 */
const SliderComponent = (
  props: Props,
  forwardedRef?: ?Ref<typeof RCTSliderNativeComponent>,
) => {
  const style = StyleSheet.compose(
    styles.slider,
    props.style,
  );

  const {
    onValueChange,
    onSlidingStart,
    onSlidingComplete,
    ...localProps
  } = props;

  const onValueChangeEvent = onValueChange
    ? (event: Event) => {
        let userEvent = true;
        if (Platform.OS === 'android') {
          // On Android there's a special flag telling us the user is
          // dragging the slider.
          userEvent =
            event.nativeEvent.fromUser != null && event.nativeEvent.fromUser;
        }
        userEvent && onValueChange(event.nativeEvent.value);
      }
    : null;

  const _disabled =
    typeof props.disabled === 'boolean'
      ? props.disabled
      : props.accessibilityState?.disabled === true;

  const _accessibilityState =
    typeof props.disabled === 'boolean'
      ? {...props.accessibilityState, disabled: props.disabled}
      : props.accessibilityState;

  const onChangeEvent = onValueChangeEvent;
  const onSlidingStartEvent = onSlidingStart
    ? (event: Event) => {
        onSlidingStart(event.nativeEvent.value);
      }
    : null;
  const onSlidingCompleteEvent = onSlidingComplete
    ? (event: Event) => {
        onSlidingComplete(event.nativeEvent.value);
      }
    : null;

  return (
    <RCTSliderNativeComponent
      {...localProps}
      thumbImage={
        Platform.OS === 'web'
          ? props.thumbImage
          : Image.resolveAssetSource(props.thumbImage)
      }
      ref={forwardedRef}
      style={style}
      onChange={onChangeEvent}
      onRNCSliderSlidingStart={onSlidingStartEvent}
      onRNCSliderSlidingComplete={onSlidingCompleteEvent}
      onRNCSliderValueChange={onValueChangeEvent}
      enabled={!_disabled}
      disabled={_disabled}
      onStartShouldSetResponder={() => true}
      onResponderTerminationRequest={() => false}
      accessibilityState={_accessibilityState}
    />
  );
};

const SliderWithRef = React.forwardRef(SliderComponent);

/* $FlowFixMe(>=0.89.0 site=react_native_fb) This comment suppresses an error
 * found when Flow v0.89 was deployed. To see the error, delete this comment
 * and run Flow. */

SliderWithRef.defaultProps = {
  value: 0,
  minimumValue: 0,
  maximumValue: 1,
  step: 0,
  inverted: false,
  tapToSeek: false,
};

let styles;
if (Platform.OS === 'ios') {
  styles = StyleSheet.create({
    slider: {
      height: 40,
    },
  });
} else {
  styles = StyleSheet.create({
    slider: {},
  });
}

/* $FlowFixMe(>=0.89.0 site=react_native_fb) This comment suppresses an error
 * found when Flow v0.89 was deployed. To see the error, delete this comment
 * and run Flow. */
const Slider = (SliderWithRef: Class<NativeComponent<Props>>);
export default Slider;
