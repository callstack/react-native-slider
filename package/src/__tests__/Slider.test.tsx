import {render, screen} from '@testing-library/react-native';
import * as React from 'react';
import renderer from 'react-test-renderer';
import Slider from '../Slider';
import {View} from 'react-native';

describe('<Slider />', () => {
  it('renders enabled slider', () => {
    render(<Slider />);

    expect(screen).toMatchSnapshot();
  });

  it('renders disabled slider', () => {
    render(<Slider disabled />);

    expect(screen).toMatchSnapshot();
  });

  it('accessibilityState disabled sets disabled={true}', () => {
    render(<Slider accessibilityState={{disabled: true}} />);

    expect(screen).toMatchSnapshot();
  });

  it('disabled prop overrides accessibilityState.disabled', () => {
    render(<Slider disabled accessibilityState={{disabled: false}} />);

    expect(screen).toMatchSnapshot();
  });

  it('disabled prop overrides accessibilityState.enabled', () => {
    render(<Slider disabled={false} accessibilityState={{disabled: true}} />);

    expect(screen).toMatchSnapshot();
  });

  it('renders a slider with custom props', () => {
    render(
      <Slider
        value={0.5}
        minimumValue={-1}
        maximumValue={2}
        step={0.25}
        minimumTrackTintColor={'blue'}
        maximumTrackTintColor={'red'}
        thumbTintColor={'green'}
        onSlidingComplete={() => {}}
        onValueChange={() => {}}
        lowerLimit={0}
        upperLimit={1}
      />,
    );

    expect(screen).toMatchSnapshot();
  });

  it('renders a slider with custom stepMaker', () => {
    render(
      <Slider
        value={2}
        minimumValue={0}
        maximumValue={4}
        StepMarker={({stepMarked}) => {
          return stepMarked ? (
            <View>
              <View style={{width: 10, backgroundColor: 'red'}} />
            </View>
          ) : (
            <View>
              <View style={{width: 10, backgroundColor: 'green'}} />
            </View>
          );
        }}
      />,
    );

    expect(screen).toMatchSnapshot();
  });

    expect(tree).toMatchSnapshot();
  });
});
