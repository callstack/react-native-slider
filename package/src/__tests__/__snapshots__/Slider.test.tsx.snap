// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider /> accessibilityState disabled sets disabled={true} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "defaultIndicatorIdle": {
          "backgroundColor": "#C0C0C0",
          "height": 10,
          "width": 2,
        },
        "defaultIndicatorMarked": {
          "backgroundColor": "#CCCCCC",
          "height": 20,
          "width": 5,
        },
        "defaultSlider": {},
        "defaultSlideriOS": {
          "height": 40,
        },
        "sliderMainContainer": {
          "width": "100%",
          "zIndex": 1,
        },
        "stepIndicatorElement": {
          "alignContent": "center",
          "alignItems": "center",
        },
        "stepNumber": {
          "alignItems": "center",
          "marginTop": 20,
          "position": "absolute",
        },
        "stepsIndicator": {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "top": 10,
          "zIndex": 2,
        },
        "thumbImage": {
          "alignContent": "center",
          "alignItems": "center",
          "position": "absolute",
        },
        "thumbImageContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
        "trackMarkContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
      },
      {
        "height": 40,
      },
      {
        "justifyContent": "center",
      },
    ]
  }
>
  <RNCSlider
    accessibilityState={
      {
        "disabled": true,
      }
    }
    disabled={true}
    inverted={false}
    lowerLimit={-9007199254740991}
    maximumValue={1}
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderAccessibilityAction={null}
    onRNCSliderSlidingComplete={null}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={0}
    style={
      [
        {
          "width": 0,
          "zIndex": 1,
        },
        {
          "height": 40,
        },
        {
          "alignContent": "center",
          "alignItems": "center",
        },
      ]
    }
    tapToSeek={false}
    thumbImage={null}
    upperLimit={9007199254740991}
  />
</View>
`;

exports[`<Slider /> disabled prop overrides accessibilityState.disabled 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "defaultIndicatorIdle": {
          "backgroundColor": "#C0C0C0",
          "height": 10,
          "width": 2,
        },
        "defaultIndicatorMarked": {
          "backgroundColor": "#CCCCCC",
          "height": 20,
          "width": 5,
        },
        "defaultSlider": {},
        "defaultSlideriOS": {
          "height": 40,
        },
        "sliderMainContainer": {
          "width": "100%",
          "zIndex": 1,
        },
        "stepIndicatorElement": {
          "alignContent": "center",
          "alignItems": "center",
        },
        "stepNumber": {
          "alignItems": "center",
          "marginTop": 20,
          "position": "absolute",
        },
        "stepsIndicator": {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "top": 10,
          "zIndex": 2,
        },
        "thumbImage": {
          "alignContent": "center",
          "alignItems": "center",
          "position": "absolute",
        },
        "thumbImageContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
        "trackMarkContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
      },
      {
        "height": 40,
      },
      {
        "justifyContent": "center",
      },
    ]
  }
>
  <RNCSlider
    accessibilityState={
      {
        "disabled": true,
      }
    }
    disabled={true}
    inverted={false}
    lowerLimit={-9007199254740991}
    maximumValue={1}
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderAccessibilityAction={null}
    onRNCSliderSlidingComplete={null}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={0}
    style={
      [
        {
          "width": 0,
          "zIndex": 1,
        },
        {
          "height": 40,
        },
        {
          "alignContent": "center",
          "alignItems": "center",
        },
      ]
    }
    tapToSeek={false}
    thumbImage={null}
    upperLimit={9007199254740991}
  />
</View>
`;

exports[`<Slider /> disabled prop overrides accessibilityState.enabled 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "defaultIndicatorIdle": {
          "backgroundColor": "#C0C0C0",
          "height": 10,
          "width": 2,
        },
        "defaultIndicatorMarked": {
          "backgroundColor": "#CCCCCC",
          "height": 20,
          "width": 5,
        },
        "defaultSlider": {},
        "defaultSlideriOS": {
          "height": 40,
        },
        "sliderMainContainer": {
          "width": "100%",
          "zIndex": 1,
        },
        "stepIndicatorElement": {
          "alignContent": "center",
          "alignItems": "center",
        },
        "stepNumber": {
          "alignItems": "center",
          "marginTop": 20,
          "position": "absolute",
        },
        "stepsIndicator": {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "top": 10,
          "zIndex": 2,
        },
        "thumbImage": {
          "alignContent": "center",
          "alignItems": "center",
          "position": "absolute",
        },
        "thumbImageContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
        "trackMarkContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
      },
      {
        "height": 40,
      },
      {
        "justifyContent": "center",
      },
    ]
  }
>
  <RNCSlider
    accessibilityState={
      {
        "disabled": false,
      }
    }
    disabled={false}
    inverted={false}
    lowerLimit={-9007199254740991}
    maximumValue={1}
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderAccessibilityAction={null}
    onRNCSliderSlidingComplete={null}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={0}
    style={
      [
        {
          "width": 0,
          "zIndex": 1,
        },
        {
          "height": 40,
        },
        {
          "alignContent": "center",
          "alignItems": "center",
        },
      ]
    }
    tapToSeek={false}
    thumbImage={null}
    upperLimit={9007199254740991}
  />
</View>
`;

exports[`<Slider /> renders a slider with custom props 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "defaultIndicatorIdle": {
          "backgroundColor": "#C0C0C0",
          "height": 10,
          "width": 2,
        },
        "defaultIndicatorMarked": {
          "backgroundColor": "#CCCCCC",
          "height": 20,
          "width": 5,
        },
        "defaultSlider": {},
        "defaultSlideriOS": {
          "height": 40,
        },
        "sliderMainContainer": {
          "width": "100%",
          "zIndex": 1,
        },
        "stepIndicatorElement": {
          "alignContent": "center",
          "alignItems": "center",
        },
        "stepNumber": {
          "alignItems": "center",
          "marginTop": 20,
          "position": "absolute",
        },
        "stepsIndicator": {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "top": 10,
          "zIndex": 2,
        },
        "thumbImage": {
          "alignContent": "center",
          "alignItems": "center",
          "position": "absolute",
        },
        "thumbImageContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
        "trackMarkContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
      },
      {
        "height": 40,
      },
      {
        "justifyContent": "center",
      },
    ]
  }
>
  <RNCSlider
    disabled={false}
    inverted={false}
    lowerLimit={0}
    maximumTrackTintColor="red"
    maximumValue={2}
    minimumTrackTintColor="blue"
    minimumValue={-1}
    onChange={[Function]}
    onRNCSliderAccessibilityAction={null}
    onRNCSliderSlidingComplete={[Function]}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={0.25}
    style={
      [
        {
          "width": 0,
          "zIndex": 1,
        },
        {
          "height": 40,
        },
        {
          "alignContent": "center",
          "alignItems": "center",
        },
      ]
    }
    tapToSeek={false}
    thumbImage={null}
    thumbTintColor="green"
    upperLimit={1}
    value={0.5}
  />
</View>
`;

exports[`<Slider /> renders disabled slider 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "defaultIndicatorIdle": {
          "backgroundColor": "#C0C0C0",
          "height": 10,
          "width": 2,
        },
        "defaultIndicatorMarked": {
          "backgroundColor": "#CCCCCC",
          "height": 20,
          "width": 5,
        },
        "defaultSlider": {},
        "defaultSlideriOS": {
          "height": 40,
        },
        "sliderMainContainer": {
          "width": "100%",
          "zIndex": 1,
        },
        "stepIndicatorElement": {
          "alignContent": "center",
          "alignItems": "center",
        },
        "stepNumber": {
          "alignItems": "center",
          "marginTop": 20,
          "position": "absolute",
        },
        "stepsIndicator": {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "top": 10,
          "zIndex": 2,
        },
        "thumbImage": {
          "alignContent": "center",
          "alignItems": "center",
          "position": "absolute",
        },
        "thumbImageContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
        "trackMarkContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
      },
      {
        "height": 40,
      },
      {
        "justifyContent": "center",
      },
    ]
  }
>
  <RNCSlider
    accessibilityState={
      {
        "disabled": true,
      }
    }
    disabled={true}
    inverted={false}
    lowerLimit={-9007199254740991}
    maximumValue={1}
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderAccessibilityAction={null}
    onRNCSliderSlidingComplete={null}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={0}
    style={
      [
        {
          "width": 0,
          "zIndex": 1,
        },
        {
          "height": 40,
        },
        {
          "alignContent": "center",
          "alignItems": "center",
        },
      ]
    }
    tapToSeek={false}
    thumbImage={null}
    upperLimit={9007199254740991}
  />
</View>
`;

exports[`<Slider /> renders enabled slider 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "defaultIndicatorIdle": {
          "backgroundColor": "#C0C0C0",
          "height": 10,
          "width": 2,
        },
        "defaultIndicatorMarked": {
          "backgroundColor": "#CCCCCC",
          "height": 20,
          "width": 5,
        },
        "defaultSlider": {},
        "defaultSlideriOS": {
          "height": 40,
        },
        "sliderMainContainer": {
          "width": "100%",
          "zIndex": 1,
        },
        "stepIndicatorElement": {
          "alignContent": "center",
          "alignItems": "center",
        },
        "stepNumber": {
          "alignItems": "center",
          "marginTop": 20,
          "position": "absolute",
        },
        "stepsIndicator": {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "top": 10,
          "zIndex": 2,
        },
        "thumbImage": {
          "alignContent": "center",
          "alignItems": "center",
          "position": "absolute",
        },
        "thumbImageContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
        "trackMarkContainer": {
          "alignContent": "center",
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
          "position": "absolute",
          "zIndex": 3,
        },
      },
      {
        "height": 40,
      },
      {
        "justifyContent": "center",
      },
    ]
  }
>
  <RNCSlider
    disabled={false}
    inverted={false}
    lowerLimit={-9007199254740991}
    maximumValue={1}
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderAccessibilityAction={null}
    onRNCSliderSlidingComplete={null}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={0}
    style={
      [
        {
          "width": 0,
          "zIndex": 1,
        },
        {
          "height": 40,
        },
        {
          "alignContent": "center",
          "alignItems": "center",
        },
      ]
    }
    tapToSeek={false}
    thumbImage={null}
    upperLimit={9007199254740991}
  />
</View>
`;
